Allow specifying a local JAR for the CUSTOM server type

FluxCapacitor2 opened the pull request:
Currently, specifying a local file path to the `BUNGEE_JAR_URL` environment variable causes an error with cURL. This PR attempts to fix that by bypassing cURL when a file path is specified instead of a URL.

Tested scenarios:
- Downloading a JAR from `https://papermc.io/api/v1/waterfall/1.16/381/download` (The script places it into a file called `download` in the `/server/` folder)
- Using a JAR in `/server/server.jar`

Fixes #34 

I am new to writing bash scripts to let me know if you see anything that needs to be changed with this! As far as I know, this won't break existing servers upgrading to this version of the script.

itzg commented on run-bungeecord.sh:
```diff
@@ -44,10 +44,10 @@ case "${TYPE^^}" in
       echo "Using custom server jar at ${BUNGEE_JAR_URL} ..."
       BUNGEE_JAR=${BUNGEE_JAR_URL}
+    elif [[ -f ${BUNGEE_JAR_URL} ]]; then
+      BUNGEE_JAR=$BUNGEE_HOME/$(basename ${BUNGEE_JAR_URL})
```
I'm now thinking either a new variable should be introduced for the file-based jar option, such as `BUNGEE_JAR_PATH` or document that if the user supplies `BUNGEE_JAR` then no further action is needed to obtain the referenced jar. 

Also rather than assume the user placed the jar in `$BUNGEE_HOME` it should just use the path as provided. For example, users of the minecraft images typically mount any number of volumes to supply pre-downloading things like this.

With that, a local variable like `$needsDownload` can be used to convey to the later part of the script if curl needs to be used or if `BUNGEE_JAR` is ready to use as is.

itzg commented on run-bungeecord.sh:
```diff
@@ -61,14 +61,15 @@ case "${TYPE^^}" in
   ;;
 esac
 
-if [ -f "$BUNGEE_JAR" ]; then
-  zarg="-z '$BUNGEE_JAR'"
-fi
```
Removing this zarg logic breaks a feature that users requested, which is to ensure that the latest bungee jar is checked even if the file is already present. I can push some follow up changes, if want to leave this fix to me.

itzg reviewed (changes requested):
Thanks for contributing this. I can push a commit with the changes I requested, if you want.

FluxCapacitor2 commented:
I have fixed most of your concerns, thanks for your feedback! Now there are two environment variables that you can use to specify a jar file location:
* `BUNGEE_JAR_URL`
* `BUNGEE_JAR_FILE`

Also, I have added the `zarg` logic back and removed the `basename` call on `$BUNGEE_JAR` so you can use jars outside of `/server`

itzg commented on run-bungeecord.sh:
```diff
@@ -43,11 +33,13 @@ case "${TYPE^^}" in
       echo "Using custom server jar at ${BUNGEE_JAR_URL} ..."
-      BUNGEE_JAR=${BUNGEE_JAR_URL}
+      BUNGEE_JAR=$BUNGEE_HOME/$(basename ${BUNGEE_JAR_URL})
+      download_required=true
```
It looks like `download_required=true` needs to be initialized as the default towards the top of the script since with all default settings, the startup fails. Here's with bash execution debugging enabled:

```
docker run -it --rm --entrypoint bash bungee
bash-4.4# bash -x /usr/bin/run-bungeecord.sh
+ : BUNGEECORD
+ : 512m
+ : 1.0.0
+ BUNGEE_HOME=/server
+ RCON_JAR_URL=https://github.com/orblazer/bungee-rcon/releases/download/v1.0.0/bungee-rcon-1.0.0.jar
+ echo 'Resolving type given BUNGEECORD'
Resolving type given BUNGEECORD
+ case "${TYPE^^}" in
+ : https://ci.md-5.net/job/BungeeCord
+ : lastStableBuild
+ : https://ci.md-5.net/job/BungeeCord/lastStableBuild/artifact/bootstrap/target/BungeeCord.jar
+ : lastStableBuild
+ BUNGEE_JAR=/server/BungeeCord-lastStableBuild.jar
+ '[' '' = true ']'
+ '[' -d /plugins ']'
+ [[ -n '' ]]
+ '[' TRUE = TRUE ']'
+ [[ ! -e /server/plugins/bungee-rcon-1.0.0.jar ]]
+ echo 'Downloading rcon plugin'
Downloading rcon plugin
+ mkdir -p /server/plugins/bungee-rcon
+ curl -sSL -o /server/plugins/bungee-rcon-1.0.0.jar https://github.com/orblazer/bungee-rcon/releases/download/v1.0.0/bungee-rcon-1.0.0.jar
+ echo 'Copy rcon configuration'
Copy rcon configuration
+ sed -i 's#${PORT}#25575#g' /tmp/rcon-config.yml
+ sed -i 's#${PASSWORD}##g' /tmp/rcon-config.yml
+ mv /tmp/rcon-config.yml /server/plugins/bungee-rcon/config.yml
+ rm -f /tmp/rcon-config.yml
+ '[' -d /config ']'
+ '[' -f /var/run/default-config.yml -a '!' -f /server/config.yml ']'
+ '[' '' = TRUE ']'
+ '[' 0 == 0 ']'
+ chown -R bungeecord:bungeecord /server
+ echo 'Setting initial memory to 512m and max to 512m'
Setting initial memory to 512m and max to 512m
+ JVM_OPTS='-Xms512m -Xmx512m '
+ '[' 0 == 0 ']'
+ exec sudo -E -u bungeecord java -Xms512m -Xmx512m -jar /server/BungeeCord-lastStableBuild.jar
Error: Unable to access jarfile /server/BungeeCord-lastStableBuild.jar
```

itzg reviewed (approved):
Thanks for the quick fixes. Looks great.

itzg commented:
The latest image now includes this change.
